src/components/tl/dpu/tl_dpu_team.c

```diff
@@ -0,0 +1,216 @@
+                                   sizeof(uint64_t), EXCHANGE_ADDR_TAG,
+                                   &send_req_param);
+    if (tl_dpu_ucp_req_check(send_req[0]) != UCC_OK) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    send_req[1] = ucp_tag_send_nbx(ctx->ucp_ep, &ctrl_seg_rkey_buf_size,
+                                   sizeof(size_t), EXCHANGE_LENGTH_TAG,
+                                   &send_req_param);
+    if (tl_dpu_ucp_req_check(send_req[1]) != UCC_OK) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    send_req[2] = ucp_tag_send_nbx(ctx->ucp_ep, ctrl_seg_rkey_buf,
+                                   ctrl_seg_rkey_buf_size, EXCHANGE_RKEY_TAG,
+                                   &send_req_param);
+    if (tl_dpu_ucp_req_check(send_req[2]) != UCC_OK) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    recv_req[0] = ucp_tag_recv_nbx(ctx->ucp_worker, rem_rkeys_lengths,
+                                sizeof(rem_rkeys_lengths),
+                                EXCHANGE_LENGTH_TAG, (uint64_t)-1,
+                                &recv_req_param);
+    if (tl_dpu_ucp_req_check(recv_req[0]) != UCC_OK) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    /* Convert to non-blocking XXX */
+    do {
+        ucp_worker_progress(ctx->ucp_worker);
+    } while((tl_dpu_ucp_req_test(&(send_req[0]), ctx->ucp_worker) != UCC_OK) ||
+            (tl_dpu_ucp_req_test(&(send_req[1]), ctx->ucp_worker) != UCC_OK) ||
+            (tl_dpu_ucp_req_test(&(send_req[2]), ctx->ucp_worker) != UCC_OK) ||
+            (tl_dpu_ucp_req_test(&(recv_req[0]), ctx->ucp_worker) != UCC_OK));
+            /* return UCC_INPROGRESS */
+
+//     fprintf (stderr,"%lu ;%lu %lu\n", rem_rkeys_lengths[0], rem_rkeys_lengths[1], rem_rkeys_lengths[2]);
+
+    ucp_rkey_buffer_release(ctrl_seg_rkey_buf);
+    total_rkey_size = rem_rkeys_lengths[0] + rem_rkeys_lengths[1] + rem_rkeys_lengths[2];
+    rem_rkeys = malloc(total_rkey_size);
+    recv_req[0] = ucp_tag_recv_nbx(ctx->ucp_worker, &rem_addresses,
+                                   sizeof(rem_addresses),
+                                   EXCHANGE_ADDR_TAG, (uint64_t)-1,
+                                   &recv_req_param);
+    if (tl_dpu_ucp_req_check(recv_req[0]) != UCC_OK) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    recv_req[1] = ucp_tag_recv_nbx(ctx->ucp_worker, rem_rkeys,
+                                total_rkey_size,
+                                EXCHANGE_RKEY_TAG, (uint64_t)-1,
+                                &recv_req_param);
+    if (tl_dpu_ucp_req_check(recv_req[1]) != UCC_OK) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    do {
+        ucp_worker_progress(ctx->ucp_worker);
+    } while((tl_dpu_ucp_req_test(&recv_req[0], ctx->ucp_worker) != UCC_OK) ||
+            (tl_dpu_ucp_req_test(&recv_req[1], ctx->ucp_worker) != UCC_OK));
+
+    self->rem_ctrl_seg = rem_addresses[0];
+    ucp_ep_rkey_unpack(ctx->ucp_ep, rem_rkeys, &self->rem_ctrl_seg_key);
```

Comment:
Why can't server allocate everything in one shot (ctr + data_in + data_out) and have 1 key for everything ? Wouldn't that  simplify the logic here? And then just use offsets into allocated shared heap for ctrl, data_in, data_out.